File: cpp/plugboard.cpp
#include "plugboard.h"
#include "utils.h"
#include <vector>
Plugboard::Plugboard(const std::string &settings)
{
	std::vector<std::string> swaps;
	int index=0;
	int last_index = 0;
	while (1)
	{
		index = settings.find(" ", last_index);
		swaps.push_back(settings.substr(last_index, index - last_index));
		last_index = index + 1;
		if (index == std::string::npos)
		{
			break;
		}
	}
	for (int i = 0; i < 26; i++)
	{
		translation_table[i] = ordalpha(i);
	}
	for (auto swap : swaps)
	{
		translation_table[alphaord(swap[0])] = swap[1];
		translation_table[alphaord(swap[1])] = swap[0];
	}
}
char Plugboard::translate_char(char c) const
{
	return translation_table[alphaord(c)];
}

File: python/plugboard.py
from utils import alphaord
from utils import ordalpha
from utils import shift_num
from utils import shift_char
from utils import construct_shifts
from utils import construct_reverse_shifts

class Plugboard:
    def __init__(self, plugboard_settings):
        swaps = plugboard_settings.split()
        self.shifts = 260 * [0]
        for pair in swaps:
            from_index = alphaord(pair[0])
            to_index = alphaord(pair[1])
            difference = to_index - from_index
            self.shifts[from_index] = difference
            self.shifts[to_index] = -difference
    def translate(self, char):
        return shift_char(char, self.shifts[alphaord(char)])

def self_test():
    plugboard = Plugboard('AZ BY')
    testvec  = 'ZYCDEFGHIJKLMNOPQRSTUVWXBA'
    alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    print 'Testing forward translation'
    result = ''
    for c in alphabet:
        result += plugboard.translate(c)
    if result == testvec:
        print 'SUCCESS' 
    else:
        print 'FAILURE'
    print 'Testing reverse translation'
    result = ''
    for c in testvec:
        result += plugboard.translate(c)
    if result ==alphabet:
        print 'SUCCESS'
    else:
        print 'FAILURE'


if __name__ == '__main__':
    self_test()
        
